cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableService.java

```diff
@@ -194,47 +209,39 @@ public void ensureTableExists(String tableName) throws IOException {
       synchronized (tables) {
         db = tables.get(tableName);
         if (db == null) {
-          createTable(tableName);
+          db = getDB(tableName, true);
+          tables.put(tableName, db);
         }
       }
     }
   }
 
-  private DB openTable(String tableName) throws IOException {
+  private DB getDB(String tableName, boolean createIfMissing) throws IOException {
     String dbPath = getDBPath(basePath, tableName);
 
     Options options = new Options();
-    options.createIfMissing(false);
+    options.createIfMissing(createIfMissing);
     options.errorIfExists(false);
     options.comparator(new KeyValueDBComparator());
     options.blockSize(blockSize);
     options.cacheSize(cacheSize);
-
-    // unfortunately, with the java version of leveldb, with createIfMissing set to false, factory.open will
-    // see that there is no table and throw an exception, but it wont clean up after itself and will leave a
-    // directory there with a lock.  So we want to avoid calling open if the path doesn't already exist and
-    // throw the exception ourselves.
-    File dbDir = new File(dbPath);
-    if (!dbDir.exists()) {
-      throw new IOException("Database " + dbPath + " does not exist and the create if missing option is disabled");
+    options.maxOpenFiles(isSystemTable(tableName) ? systemTablesMaxOpenFiles : userTablesMaxOpenFiles);
+
+    if (!createIfMissing) {
+      // unfortunately, with the java version of leveldb, with createIfMissing set to false, factory.open will
+      // see that there is no table and throw an exception, but it wont clean up after itself and will leave a
+      // directory there with a lock.  So we want to avoid calling open if the path doesn't already exist and
+      // throw the exception ourselves.
+      File dbDir = new File(dbPath);
+      if (!dbDir.exists()) {
+        throw new IOException("Database " + dbPath + " does not exist and the create if missing option is disabled");
```

Comment:
I feel that this should be a NotFoundException so that the caller can handle it appropriately, rather than a generic IOException.

Reply:
Most of the usages of this method are from `LevelDBTableCore`, at which point this exception won't be thrown, unless someone explicitly deletes the directory for the db.
I don't think `LevelDBTable` can meaningfully handle a `NotFoundException` either, differently than any other `IOException` it receives at that point.

Reply:
Agreed but in general it is a better habit to throw meaningful exceptions - or do you know what other classes will use this in the future? Or who will copy your code into their own?